accounts/README.md

```diff
@@ -1,10 +1,123 @@
+It is RECOMMENDED that the account supports the DELEGATECALL operation.
+
+#### Hooks
+
+Hooks do not require any specific interface to be implemented by the account. To use hooks, use the interface defined in the
+[Safe{Core} Protocol Manager specification](../manager/README.md)
+
+#### Function Handlers
+
+The account MUST implement a `fallback` function, which forwards all the calldata
+to the Safe{Core} Protocol Manager AND appends the 20 bytes of the `msg.sender` address to the calldata.
+
+An example implementation of such a `fallback` function is:
+```Solidity
+    fallback() external {
+        bytes32 slot = SAFE_CORE_PROTOCOL_MANAGER_SLOT;
+        assembly {
+            function allocate(length) -> pos {
+                pos := mload(0x40)
+                mstore(0x40, add(pos, length))
+            }
+
+            let protocol := sload(slot)
+            if iszero(protocol) {
+                return(0, 0)
+            }
+
+            let calldataPtr := allocate(calldatasize())
+            calldatacopy(calldataPtr, 0, calldatasize())
+
+            // The msg.sender address is shifted to the left by 12 bytes to remove the padding
+            // Then the address without padding is stored right after the calldata
+            let senderPtr := allocate(20)
+            mstore(senderPtr, shl(96, caller()))
+
+            // Add 20 bytes for the address appended add the end
+            let success := call(gas(), protocol, 0, calldataPtr, add(calldatasize(), 20), 0, 0)
+
+            let returnDataPtr := allocate(returndatasize())
+            returndatacopy(returnDataPtr, 0, returndatasize())
+            if iszero(success) {
+                revert(returnDataPtr, returndatasize())
+            }
+            return(returnDataPtr, returndatasize())
+        }
+    }
+```
+
+#### Signature Validators
+
+1. The account MUST follow the requirements defined in the Function Handlers section, or the account MUST forward the EIP-1271 isValidSignature call to the Safe{Core} Validator Manager, following the expected encoding defined in the [Signature Validator specs](../modules/README.md).
+2. The account MUST implement the following interface:
+
+```Solidity
+interface IAccount {
+    function domainSeparator() external view returns (bytes32);
+
+    function checkSignatures(bytes32 messageHash, bytes memory messageData, bytes memory signatures) external view;
+}
+```
+where:
+- `domainSeparator()` function MUST return the domain separator hash following the [EIP-712](https://eips.ethereum.org/EIPS/eip-712#definition-of-domainseparator) standard
+- `checkSignatures` function parameters are:
+
+  | Type         | Name        | Description            |
+  |--------------|-------------|------------------------|
+  | bytes32      | messageHash | Hash of the message    |
+  | bytes memory | messageData | EIP-712 pre-image data |
```

Comment:
Is this always available? I'm asking because I feel like I've seen calls to this function with `messageData = ""` and can imagine scenarios where this isn't available (for example, when verifying EIP-1271 signatures for example with the default verifier).

Reply:
So the `messageData` variable has a legacy background because the older EIP-1271 spec used `bytes` instead of `bytes32` and the data hash. If it's not always available, I'd explicitly state that the account must not expect this argument but keep the signature consistent

Reply:
Looking through the `checkSignatures` code, it looks like older versions of the Safe used this for the old EIP-1271 spec for contract signatures. Seeing as, since Safe v1.5.0 this is no longer the case (it uses the the new EIP-1271 signature verification for all contract signatures), could we just remove `data` from the interface?

If the old `checkSignatures` function is kept around for backwards compatibility on the Safe, that is fine, but I don't think it should be needed for the `IAccount` interface. Essentially, you can add the following methods to the Safe so keep a backwards compatible interface, without adding the `data` "tech debt" to the new `IAccount` interface:

```solidity
function checkSignatures(bytes32 dataHash, bytes memory /* data */, bytes memory signatures) public view {
    checkSignatures(dataHash, signatures);
}

function checkNSignatures(
    address executor,
    bytes32 dataHash,
    bytes memory /* data */,
    bytes memory signatures,
    uint256 requiredSignatures
) public view {
    checkNSignatures(executor, dataHash, signatures, requiredSignatures);
}
```

This can even be implemented as a plugin to not take precious code space from the `Safe` contract.